### ISSUE TITLE
TypeError Cannot read property 'body' of null

### ISSUE THREAD

**Issue by AvraamMavridis** [testing-library/react-testing-library#661]
- `@testing-library/react` version: 10.0.4
- `react` version: 16.13.1
- `node` version: 11
- `npm` (or `yarn`) version: 6.14.5

### What happened:

Cleanup is happening before react finishes its work, crashing the test

### Reproduction:

https://repl.it/repls/PeskyStimulatingBetaversion

### Problem description:

![Screenshot 2020-05-07 at 11 23 04](https://user-images.githubusercontent.com/283690/81278006-6a675c00-9055-11ea-9530-4b0ec97cf80b.png)

**Comment by AvraamMavridis** [comment]
Related https://github.com/facebook/react/issues/15691

**Comment by kentcdodds** [comment]
It's possible that this is a bug in the coverage instrumentation transformation. Do you get this error only when running with `--coverage` enabled?

**Comment by AvraamMavridis** [comment]
@kentcdodds its without --coverage. Here is my command

`cross-env CI=true jest --verbose --color`

The component where it crashes is looking like:

```js
const ShippedIngredientsProvider = function ShippedIngredientsProvider(props) {
  const { selectedCountry } = props;
 
  const [shippedIngredientsState, setShippedIngredientsState] = useState({
    ingredients: [],
  });

  let updateState = true;

  const setState = obj => {
    if (updateState) {
      // crashes here
      setShippedIngredientsState({ ...shippedIngredientsState, ...obj });
    }
  };

  useEffect(() => {
    // ... some fetching happens here

    return function cleanUp() {
      updateState = false;
    };
  }, [selectedCountry]);

```

**Comment by kentcdodds** [comment]
I think you missed some of that component... The code you have right there isn't returning anything.

**Comment by AvraamMavridis** [comment]
@kentcdodds sorry updated

**Comment by kentcdodds** [comment]
I'm this case I believe the solution would be to use `waitFor` to make sure your test waits to complete until after the UI had settled.

Cleanup happens when the test is finished and you're not waiting for the component to do it's work before the test finishes. Good luck!

**Comment by derekprice** [comment]
@kentcdodds Could you explain a little further what you meant when you mentioned the coverage instrumentation transformation potentially being an issue? We're still running into this issue (passing locally, failing _only sometimes_ in CI when using --coverage) and we're using `waitFor` for each assertion in these tests.

**Comment by kentcdodds** [comment]
Sometimes the code that's generated for coverage can have issues that change the behavior of the software. That would be a bug in the Babel plugin responsible for that.

**Comment by derekprice** [comment]
@kentcdodds Got it, that makes sense. 

After doing more investigation, it actually looks like 90% of the time that our RTL tests are failing with the `'body' of null` error, it's referencing an unhandled API failure in a separate component file, which triggers a [sweetalert](https://github.com/t4t5/sweetalert) popup that seems to be causing issues. (For context, we just started incorporating RTL in our test suite which is mostly Enzyme.)

I know that `cleanup` gets run automatically if you use Jest, but do you think running `cleanup` _before_ each test here could help to account for things like this? 

For example:
```
  ● Settings › blocks the user from accessing majority of settings if criteria aren't met

    TypeError: Cannot read property 'body' of null

      12 | 
      13 |   const errMsg = errorList && errorList[0] ? errorList[0].message : error;
    > 14 |   swal('Sorry! Something went wrong', errMsg, 'error');
         |   ^
      15 | };
      16 | 
      17 | export const setErrorMessage = error => {

      at swal (node_modules/sweetalert/lib/sweetalert.js:64:157)
      at handleApiError (frontend/util/errors.js:14:3)
      at Promise.all.then.catch.err (frontend/plans/plan-status.jsx:61:34)
```

**Comment by kentcdodds** [comment]
If you can come up with a small reproduction following [the instructions here](https://kcd.im/rtl-help), then the community can help you out :)

**Comment by ilovett** [comment]
Check if you are on latest `@testing-library/*` packages and also using `jest-environment-jsdom-sixteen` fixed this issue for me.

**Comment by kimskovhusandersen** [comment]
Hi everyone,

I encountered a similar issue where my tests were not behaving as expected. After some investigation, I found that the solution was to wrap my events in an await act.

Here’s an example of how I adjusted my code:

Before:

```
fireEvent.click(button);
expect(something).toBe(someValue);

```
After:
```
await act(async () => {
  fireEvent.click(button);
});
expect(something).toBe(someValue);
```
This ensures that all updates related to the event have been processed before jest is torn down.

I hope it helps others facing the same issue!
